### ISSUE TITLE
Need assistance with notebook experiment check_zero_shot_results

### ISSUE THREAD

**Issue by bgoldfe2** [stevebottos/owl-vit-object-detection#11]
@stevebottos Thanks for the really great code! I am learning a lot. Quick issue, I am not able to get the check_zero_shot_results.ipynb working (it may be based on some older versions of your code). Can you please update as there are some mismatches with file locations and parameters that I am unable to resolve. Thanks!

**Comment by stevebottos** [comment]
Yep, you're right - it's out of date. I plan on making some updates tonight as well as a bit of a refactor. Should be good to go by tomorrow!

**Comment by stevebottos** [comment]
I haven't had a minute to get to it yet. I've been preoccupied trying to solve some strange training issues. Have you tried training the model on your dataset yet?

**Comment by bgoldfe2** [comment]
I have been doing some test training with COCO successfully. Just checking if it was able to run the epochs. I was wondering about where you got your cost function from which adds up all the costs.

**Comment by bgoldfe2** [comment]
I am reformatting other data sets with a different viewpoint such as the Kaggle Ship Detection dataset to see if I could fine-tune the pretrained owl-vit from natural view to top looking down view effectively. I should have some numbers next week on how well that works.

**Comment by stevebottos** [comment]
@bgoldfe2 The cost function I just made myself, I figured that if owl was originally trained contrastively then it would be best to continue training it contrastively (the idea of the loss function is to enforce high cosine similarity between similar embeddings and push non similar ones away). It could be the cause of the issues, but I'm not sure.

If you could post some numbers that would be great. I've been a bit busy so I haven't been able to make any code updates but I'm thinking I'll collect some numbers too soon and post benchmarks.

**Comment by bgoldfe2** [comment]
@stevebottos I was able to do some conversions to the data set I am using to put it into the COCO format and was able to run training for two epochs just for an "is alive" test. The results are not good, but there is definite improvement from epoch 1 to epoch 2. I will do some more epochs and see what happens and post more.
 epoch  class loss (T/V)    bg loss (T/V)      map    map@0.5  map (L/M/S)     mar (L/M/S)     time elapsed
-------  ------------------  ---------------  -----  ---------  --------------  --------------  --------------
      0  3.8302/3.8161       1.2377/1.2321    0.051      0.099  0.1/0.09/0.01   0.25/0.16/0.03  0:07:49
      1  3.3322/3.3249       1.1454/1.1417    0.072      0.131  0.13/0.11/0.02  0.32/0.18/0.04  0:15:40

**Comment by stevebottos** [comment]
Interesting, I wouldn't expect the results to be that poor. What numbers are you getting with coco data?

**Comment by stevebottos** [comment]
btw, I was tinkering last night and I might include some alternative training methods - a more traditional approach of just slapping an MLP classification head and using cross entropy on the outputs might be better in some cases (similar to using a linear probe on top of CLIP).

**Comment by bgoldfe2** [comment]
This is what I got when I ran a a subset of COCO ~10k images through the other week. It did not match your results but 10x better than the aerial view results. 
  epoch  class loss (T/V)    bg loss (T/V)      map    map@0.5  map (L/M/S)     mar (L/M/S)     time elapsed
-------  ------------------  ---------------  -----  ---------  --------------  --------------  --------------
      0  0.4315/0.4295       0.2336/0.2316    0.318      0.507  0.5/0.3/0.14    0.65/0.46/0.21  0:07:48
      1  0.4107/0.4096       0.2103/0.2097    0.321      0.512  0.51/0.31/0.14  0.65/0.46/0.21  0:15:34
      2  0.3984/0.3978       0.2006/0.2002    0.324      0.518  0.51/0.31/0.14  0.66/0.46/0.21  0:23:17
      3  0.3896/0.3891       0.1946/0.1944    0.324      0.516  0.51/0.31/0.14  0.66/0.45/0.21  0:30:57
      4  0.3827/0.3823       0.1906/0.1904    0.322      0.515  0.51/0.3/0.14   0.66/0.45/0.21  0:38:41
      5  0.377/0.3767        0.1875/0.1874    0.322      0.516  0.51/0.31/0.14  0.66/0.46/0.21  0:46:28
      6  0.3722/0.372        0.1851/0.185     0.325      0.521  0.51/0.31/0.14  0.66/0.45/0.21  0:54:22
      7  0.368/0.3678        0.183/0.183      0.322      0.518  0.51/0.3/0.14   0.66/0.45/0.21  1:02:11
      8  0.3643/0.3641       0.1814/0.1813    0.324      0.517  0.51/0.3/0.14   0.66/0.45/0.22  1:09:58
      9  0.361/0.3609        0.1799/0.1799    0.324      0.517  0.51/0.31/0.14  0.66/0.45/0.22  1:17:42
     10  0.358/0.3579        0.1787/0.1786    0.323      0.516  0.51/0.31/0.14  0.66/0.45/0.22  1:25:29
     11  0.3554/0.3553       0.1776/0.1775    0.326      0.522  0.51/0.31/0.14  0.66/0.45/0.22  1:33:18
     12  0.3529/0.3528       0.1765/0.1765    0.323      0.518  0.51/0.3/0.14   0.66/0.45/0.22  1:41:00
     13  0.3506/0.3505       0.1756/0.1756    0.322      0.514  0.51/0.29/0.15  0.66/0.44/0.22  1:48:50
     14  0.3484/0.3484       0.1748/0.1748    0.318      0.513  0.51/0.29/0.14  0.65/0.44/0.21  1:56:40
     15  0.3465/0.3464       0.1741/0.174     0.319      0.511  0.51/0.29/0.14  0.66/0.44/0.22  2:04:20
     16  0.3446/0.3445       0.1734/0.1734    0.318      0.514  0.51/0.29/0.14  0.66/0.44/0.22  2:12:03
     17  0.3428/0.3428       0.1728/0.1727    0.319      0.513  0.51/0.29/0.14  0.66/0.44/0.22  2:19:51
     18  0.3412/0.3411       0.1722/0.1721    0.32       0.512  0.51/0.29/0.14  0.66/0.44/0.22  2:27:40
     19  0.3396/0.3396       0.1716/0.1716    0.321      0.517  0.5/0.29/0.14   0.65/0.44/0.22  2:35:31

**Comment by bgoldfe2** [comment]
I was thinking similar with a cross-entropy loss for backprop. If we are using the same loss function as pre-training,  aren't we just doing more pre-training, not fine-tuning? I have not experimented with just using an MLP head, but would a a two-layer classification network with GELU provide us with valid output? Also, is there a way with your code to visualize the bbox and image ouputs? Thanks.

**Comment by bgoldfe2** [comment]
I am going to do some reading in the recent survey paper, "Vision-Language Models for Vision Tasks: A Survey" (https://arxiv.org/pdf/2304.00685.pdf). I am hoping that will provide some insight. It may be for the data I am using, vertical view versus natural view, I may have to start blank and pre-train a new model. Thoughts?

**Comment by stevebottos** [comment]
I have a local branch that implements what you described, and it actually performs worse than what's in main. I don't believe there's an issue with using the same loss function during fine tuning as in pretraining, that's pretty common actually. The idea of the contrastive loss is to allow for better generalization in few shot scenarios since contrastive loss with cosine classifiers have shown to do well in that task.

**Comment by bgoldfe2** [comment]
Thanks so much for the advisement! How would I get some output for the test images to be able to visualize the label and bounding boxes per image? Thanks.

**Comment by stevebottos** [comment]
You should be able to modify the config, there' a field `save_eval_images` that you can set to `true`. Just ensure that you bump up the confidence/nms threshold to something reasonable first (0.75/0.5 is what I'd start with). The super low values are just used for benchmarking.

**Comment by bgoldfe2** [comment]
Hi @stevebottos I am working with your updated code from 3 weeks ago and COCO training using default setting is running a bit better, but not as good as your original results. Here is output from 5 epochs:
  epoch  class loss (T/V)    bg loss (T/V)      map    map@0.5  map (L/M/S)     mar (L/M/S)     time elapsed
-------  ------------------  ---------------  -----  ---------  --------------  --------------  --------------
      0  3.3374/3.318        0.837/0.8288     0.346      0.534  0.5/0.31/0.13   0.63/0.43/0.18  0:01:58
      1  3.0165/3.0067       0.776/0.7718     0.352      0.538  0.51/0.31/0.12  0.64/0.43/0.17  0:03:53
      2  2.8381/2.8308       0.7388/0.7362    0.35       0.535  0.52/0.31/0.12  0.64/0.42/0.18  0:05:49
      3  2.7147/2.7104       0.7143/0.712     0.352      0.536  0.52/0.3/0.12   0.65/0.42/0.17  0:07:46
      4  2.6263/2.6232       0.6958/0.694     0.353      0.536  0.52/0.31/0.12  0.64/0.42/0.18  0:09:41

**Comment by stevebottos** [comment]
Good call, this commit: `f47d69015f00472f8709f7e6a9149f20a3d5807d` is the best state that the repo has been in, I'll be reverting to that soon and cleaning up the code/providing inference scripts and whatnot.

Your results look reasonable to me, the script provided will take a random subset from coco so yours is different than the one that was benchmarked on. I wouldn't worry about it. I'll try to do some experimenting after refactoring/reverting that will hopefully be able to squeeze out some more accuracy.

**Comment by bgoldfe2** [comment]
I am looking at the output which is a bit different from the inputs :-) Are there techniques you know of to mitigate the hyper-generation of bbox's:
![image](https://github.com/stevebottos/owl-vit-object-detection/assets/23743490/ed256957-fe00-44a3-8431-76ee1334c42c)

**Comment by stevebottos** [comment]
You need to up the confidence and iou thresholds in `config.yaml` to inference numbers, I'd play around with them but definitely raise them from the 0.001/0.6 used for benchmarking. Benchmarking is done typically with these numbers, since under the hood coco will remove boxes. In practice though we don't use coco to evaluate and need to raise the values I mentioned. Expect a drop in accuracy from a numbers standpoint (precision and recall will drop since you're retrieving only boxes you're sure of), but you will notice that the result makes more sense visually.

**Comment by bgoldfe2** [comment]
@stevebottos Thanks, much better visual output! I will need to play with the settings.
![image](https://github.com/stevebottos/owl-vit-object-detection/assets/23743490/8836a77f-4056-4faa-9084-86c1c957dfb9)

**Comment by bgoldfe2** [comment]
@stevebottos I am also noticing that the weight decay in the config.yaml is not impacting the training. It is always 0.1
[3e-05]
100%|███████████████████| 2481/2481 [01:46<00:00, 23.35it/s]
100%|█████████████████████| 200/200 [00:24<00:00,  8.03it/s]
Computing metrics...

  epoch  class loss (T/V)    bg loss (T/V)      map    map@0.5  map (L/M/S)     mar (L/M/S)    time elapsed
-------  ------------------  ---------------  -----  ---------  --------------  -------------  --------------
      0  2.4487/2.4286       0.675/0.6646     0.093      0.131  0.14/0.09/0.02  0.15/0.1/0.02  0:02:15

[2.7000000000000002e-05]

**Comment by bgoldfe2** [comment]
@stevebottos is there a way to save the model after fine-tuning? Thanks.

**Comment by bgoldfe2** [comment]
@stevebottos would you be open to me providing some code into this effort? Right now I have a fork of just main. I have some local changes that I would like to add more on the managing different Runs and having a folder with the config for the run, hopefully the model, and any outputs (plots, images w/bbox, results). I think I can do pull requests from a fork but have never done that. Thoughts?

**Comment by stevebottos** [comment]
The model is just a pytorch model so saving as described here is what you are looking for https://pytorch.org/tutorials/beginner/saving_loading_models.html 

I would be happy to have you contribute, that sounds great. I can grant you access and you can open up a PR in this repo cherrypicking your changes over. I'm planning on refactoring the code soon in any case to provide more functionality and easier use, namely larger batch sizes and packaging this as a library to make it easier to interact with.

**Comment by bgoldfe2** [comment]
@stevebottos, I see you reverted back to the July 8 commit for main, f47d69015f00472f8709f7e6a9149f20a3d5807d. I will try running my data using that version of the code. I see that there is a change in the loss function. Can you explain the Push/Pull loss function and a reference to where you found that or how you came up with that? In the July 8 version the neg_loss subtracts the pos_loss and in the one I forked it is the neg_loss? Thanks.

neg_loss = (torch.pow(1 - torch.exp(-neg_loss), 2) * _**pos_loss**_).sum(
            dim=0
        ) / self.n_classes

**Comment by stevebottos** [comment]
Yep, that is just focal loss essentially. That is definitely a mistake there, it should be neg_loss.

The general idea of push/pull loss is that since we're contrastively learning an optimal set of vectors to use as priors to get similarities, we want to pull new embeddings of the same class as close to that class's prior vector as possible while pushing others away.

By the way, I found two problems with the code. I adopted DETR's loss function, but I'm not using DETR's transformer decoder after the encoder. This causes issues because what OWL-ViT predicts are many boxes, many of which are more or less around the same object (often 5+ with high-ish iou around the object of interest). This causes problems during training because we're only picking one of many possibly valid bounding boxes around an object, and the model gets confused and fails to optimize.

I'm working on a fix in which additional bounding boxes are labeled as ground truths, I'd also like to attach DETR's decoder to see if I can improve the accuracy. I'll push the code soon, it's in a broken state right now and I have no idea why. I ran a training session overnight with 10k images and this is what I came back to lol:

```  epoch    class loss    bg loss    box loss    map    map@0.5  map (L/M/S)     mar (L/M/S)     time elapsed
-------  ------------  ---------  ----------  -----  ---------  --------------  --------------  --------------
      0      11.6731     0.76779     0.28675  0.202      0.351  0.32/0.17/0.05  0.55/0.38/0.13  0:16:21
      1       7.81454    0.6436      0.28437  0.222      0.381  0.33/0.21/0.06  0.56/0.39/0.14  0:32:35
      2       6.27058    0.58497     0.28266  0.231      0.39   0.35/0.2/0.07   0.57/0.41/0.16  0:48:09
      3       5.33311    0.54838     0.2813   0.237      0.402  0.36/0.21/0.07  0.56/0.41/0.15  1:03:47
      4       4.70999    0.52327     0.28012  0.244      0.419  0.36/0.22/0.07  0.56/0.4/0.16   1:20:33
      5       4.23129    0.50459     0.27914  0.257      0.44   0.39/0.22/0.06  0.57/0.41/0.15  1:37:05
      6       3.86843    0.48939     0.27818  0.248      0.427  0.38/0.23/0.07  0.58/0.41/0.17  1:54:13
      7       3.55938    0.47563     0.27741  0.255      0.435  0.39/0.23/0.06  0.57/0.4/0.15   2:07:45
      8       3.29837    0.46266     0.27665  0.269      0.448  0.41/0.24/0.07  0.58/0.42/0.17  2:20:19
      9       3.08629    0.4529      0.27578  0.261      0.443  0.39/0.23/0.07  0.59/0.4/0.15   2:32:53
     10       2.90684    0.44374     0.27505  0.271      0.456  0.41/0.23/0.07  0.58/0.42/0.17  2:45:23
     11      71.569      0.55585     0.29781  0          0      0.0/0.0/0.0     0.0/0.0/0.0     2:55:35
     12     143.28       0.57612     0.31022  0          0      0.0/0.0/0.0     0.0/0.0/0.0     3:06:06
     13     204.599      0.56451     0.31659  0          0      0.0/0.0/0.0     0.0/0.0/0.0     3:17:06
     14     257.681      0.54681     0.32023  0          0      0.0/0.0/0.0     0.01/0.0/0.0    3:27:55
     15     304.041      0.52928     0.32284  0          0      0.0/0.0/0.0     0.0/0.0/0.0     3:38:37
     16     344.986      0.51096     0.32478  0          0      0.0/0.0/0.0     0.0/0.0/0.0     3:49:26
     17     381.369      0.49346     0.32607  0          0      0.0/0.0/0.0     0.0/0.0/0.0     3:59:54
     18     413.92       0.47695     0.32699  0          0      0.0/0.0/0.0     0.01/0.0/0.0    4:10:20
     19     443.212      0.46152     0.32757  0          0      0.0/0.0/0.0     0.01/0.0/0.0    4:20:38
     ```
     
I'll need to investigate why the gradient exploded, but it looks like it's learning much better and would have continued improving

**Comment by bgoldfe2** [comment]
Concur, training looks good until it explodes.

**Comment by stevebottos** [comment]
Yep, tonight I'll try to clip the gradients and see if that helps. In any case, training in main right now is broken (it'll work for really short training runs 1-3 epochs) because it does not implement the soft-positive labeling of bounding boxes with very high overlap from the ground truth. Once I've got that fixed it'll be ready to cleanup/refactor and release. I'll give you access after I get that refactor done so you can start opening up PR's too.

**Comment by stevebottos** [comment]
Go ahead and give the latest push a try, it has a bunch of new features

**Comment by aidyai** [comment]
I have been following this and I so much appreciate the efforts you both are putting in.

I want to know are you using a 24gb GPU or Multi GPU training and are you using the original Jax implementation as a guide?

**Comment by bgoldfe2** [comment]
@aidyai I am running it on a single p100 gpu with 32GB for the dataset I am using it takes about 2 minutes per epoch to run. I also have a system with a consumer RTX 3090 w/ 24GB I will try next week. One of the adaptations we can make would be to run using the Hugging Face Accelerator library that looks easy to implement, That can optimize even on a single GPU according to the documentation.

**Comment by stevebottos** [comment]
I'm not using the Jax implementation from the original google research repo as a guide, I'm going off of the huggingface implementation since I like pure pytorch instead of jax. 

As far as acceleration goes @bgoldfe2 , if they accelerate by using mixed precision it won't work for this unfortunately since I'm usine BCELoss in the loss function which can't be cast down to half precision. I can't switch to BCEWithLogitsLoss (which is numerically stable at half precision) because that automatically applies a sigmoid to the input of the loss whereas the "logits" I use in the loss aren't actually logits at all but rather similarities that are already between 0-1. However, I don't know if Huggingface's accelerator uses mixed precision, I'd assume it does though.

An option would be to disable mixed precision in the loss function so that there are still speedups gained by autocasting the model to mixed. I can look into that. But, I want to first look into implementing a more generalized version of CLIP's contrastive loss function for object detection instead of BCELoss.

**Comment by aidyai** [comment]
Honestly, I really love this implementation, I am trying to create a dataset that performs object detection on UI Designs, as such I want to inquire if it will be worth it to use this implementation considering this the nature of what I want to achieve is related an Open Vocabulary kind of detection, then if I am using the code for finetuning, what do you think will be the minimum optimal dataset size because I saw something like you are working with a dataset size of over 10k.?

**Comment by stevebottos** [comment]
@aidyai I think that for open vocabulary detection you would want to use the out-of-the-box owl-vit, this implenetation strips the text portion from the model entirely in order to switch objectives from general open-vocabulary object detection to strict object detection (no text prompts involved). 

I've also found that the current implementation works best with few examples and the larger the dataset grows the worse the performance is, the use case would be if you have very minimal data and you still want to obtain decent results. For open vocabulary stuff I believe you'd need a massive dataset. In any case I do plan to maybe include an implementation for open-vocabulary fine tuning in the future, but right now my main concern is improving object detection results. I'm hoping to have a working update some time soon.

**Comment by aidyai** [comment]
@stevebottos Thanks so much, really appreciate
